### PR TITLE
Implement specification pattern with new endpoints

### LINKED ISSUE #246: New endpoints for IR-API for the frontend
Based on this: https://github.com/interactivereduction/frontend/pull/36#issuecomment-1900064341

### LINKED ISSUE #270: Refactor IR-API to improve query optimizations
Allow easier migration to GraphQL in the future, and solve some issues with performance now.

### PR DESCRIPTION
**Pull request by keiranjprice101**
closes #270 
closes #246 
closes #252

This PR does 3 things:
 - Implements the specifications with the repo pattern
 - Adds the new endpoints for the frontend
 - Improves API docs page

The spec and repo, results in a significantly reduced test burden and other benefits discussed. As well as allowing us to more cleanly apply sorting and pagination via simple function and decorator.

For the frontend endpoint changes

It is now possible to filter by run fields that exist on the frontend table from the /instrument/<>/reductions table
The count endpoints have been added
And the endpoint to see all runs for an instrument

Based on now closed #251

### CODE REVIEW

File: ir_api/core/specifications/reduction.py [hunk 1]
```diff
@@ -0,0 +1,117 @@
+"""
+Module defining specifications for querying Reduction entities within the IR API.
+
+It includes the ReductionSpecification class, which facilitates the construction of complex queries
+for fetching Reduction entities based on various criteria such as instrument name, experiment number,
+and ordering preferences.
+"""
+# pylint: disable=unused-argument
```
Comment (keiranjprice101) [inline ir_api/core/specifications/reduction.py:8]: See: https://github.com/interactivereduction/IR-API/pull/277/files#diff-c84d21f7021977eeefaca8726ed64293b40713a2ad0b5acfa6afc78fe7841072R64 for disable reason

File: ir_api/core/specifications/run.py [hunk 2]
```diff
@@ -0,0 +1,54 @@
+"""
+Module defining specifications for querying Run entities in the IR API.
+
+It includes the RunSpecification class which allows for building complex queries
+for Run entities based on various criteria such as instrument, limit, offset,
+order by, and order direction.
+"""
+# pylint: disable=unused-argument
```
Comment (keiranjprice101) [inline ir_api/core/specifications/run.py:8]: See: https://github.com/interactivereduction/IR-API/pull/277/files#diff-c84d21f7021977eeefaca8726ed64293b40713a2ad0b5acfa6afc78fe7841072R64
for disable reason

**Review by Dagonite — APPROVED** [review]
Looks good to me!

**Comment by keiranjprice101** [comment]
@Pasarus reready

**Review by Pasarus — CHANGES_REQUESTED** [review]
Just 2 small things

**Review by Pasarus — APPROVED** [review]

**Pull request merged**